waku/v2/waku_core/topics/sharding.nim

```diff
@@ -0,0 +1,113 @@
+## Waku autosharding utils
+##
+## See 51/WAKU2-RELAY-SHARDING RFC: https://rfc.vac.dev/spec/51/#automatic-sharding
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  nimcrypto,
+  std/math,
+  std/sequtils,
+  std/algorithm,
+  std/strutils,
+  stew/endians2,
+  stew/results,
+  stew/byteutils
+
+import
+  ./content_topic,
+  ./pubsub_topic
+
+const ClusterIndex* = 49152
+const GenerationZeroShardsCount* = 5
+
+type ShardsPriority = seq[tuple[topic: NsPubsubTopic, value: float64]]
+
+proc applyWeight(hashValue: uint64, weight: float64): float64 =
+  -weight / math.ln(float64(hashValue) / float64(high(uint64)))
+
+proc hashOrder*(x, y: (NsPubsubTopic, float64)): int =
+    cmp(x[1], y[1])
+
+proc weightedShardList*(topic: NsContentTopic, shardCount: int, weights: seq[float64]): Result[ShardsPriority, string] =
+  ## Returns the ordered list of shards and their priority values.
+
+  if weights.len != shardCount:
+    return err("Must provide weights for every shards")
+
+  let shardsNWeights = zip(toSeq(0..shardCount), weights)
+
+  var list = newSeq[(NsPubsubTopic, float64)](shardCount)
+
+  for (shard, weight) in shardsNWeights:
+    let pubsub = NsPubsubTopic.staticSharding(ClusterIndex, uint16(shard))
+    let bytes = toBytes($topic) & toBytes($pubsub)
+    let hash = sha256.digest(bytes)
+    let hashValue = uint64.fromBytesBE(hash.data)
+    let value = applyWeight(hashValue, weight)
+
+    list[shard] = (pubsub, value)
+
+  list.sort(hashOrder)
+
+  ok(list)
+
+type ShardingBias* = enum
+  None = "none"
+  Kanonymity = "anon"
+  Throughput = "bandwidth"
+
+proc shardingParam*(topic: NsContentTopic): Result[(int, ShardingBias), string] =
+  ## Returns the total shard count and the sharding selection bias
+  ## from the content topic.
+  let gen = try:
+    parseInt(topic.generation)
+  except ValueError:
+    return err("Cannot parse generation: " & getCurrentExceptionMsg())
+
+  let shardCount =
+    case gen:
+      of 0:
+        GenerationZeroShardsCount
+      else:
+        return err("Generation > 0 are not supported yet")
+
+  let bias = try:
+    parseEnum[ShardingBias](topic.bias)
+  except ValueError:
+    return err("Cannot parse sharding bias: " & getCurrentExceptionMsg())
+
+  ok((shardCount, bias))
+
+proc biasedWeights*(shardCount: int, bias: ShardingBias): seq[float64] =
+  var weights = repeat(1.0, shardCount)
+
+  case bias:
+    of None:
+      return weights
+    of Kanonymity:
+      # we choose the lower 20% of shards and double their weigths
+      let index = shardCount div 5
+      for i in (0..<index):
+        weights[i] *= 2.0
+    of Throughput:
```

Comment:
Is there any "assurance" that the higher shards will have better bandwidth, other than our assumption that the lower shards may have more users?

Reply:
We have no direct control, some "anon" biased shards MAY end up less populated than "bandwidth" biased ones.

Even without, a way to bias the shard selection it's the same.

Reply:
Yeah, I understand we would have the same problem without the bias, but the bias (and the naming) kinda implies to me that we are offering a solution to the problem. Just wanted to point it out - might be important to mention/explain in docs?

Reply:
I renamed the bias to what they actually do.